compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp

```diff
@@ -0,0 +1,537 @@
+  return llvm::to_vector(
+      llvm::map_range(interchange, [&](unsigned val) { return vector[val]; }));
+}
+/// Helper method to apply to invert a permutation.
+static SmallVector<unsigned> invertPermutationVector(
+    ArrayRef<unsigned> interchange) {
+  SmallVector<unsigned> inversion(interchange.size());
+  for (auto pos : llvm::enumerate(interchange)) {
+    inversion[pos.value()] = pos.index();
+  }
+  return inversion;
+}
+/// Method to check if an interchange vector is a permutation.
+static bool isPermutation(ArrayRef<unsigned> interchange) {
+  llvm::SmallDenseSet<unsigned, 4> seenVals;
+  for (auto val : interchange) {
+    if (seenVals.count(val)) return false;
+    seenVals.insert(val);
+  }
+  return seenVals.size() == interchange.size();
+}
+
+/// Given the `lb` and `step` of a loop, return the lower bound and step to use
+/// for a distributed loop. Replace the iteration domain to
+/// - lb_partitioned = lb + procId * step
+/// - step_partitioned = step * nprocs
+static std::tuple<Value, Value> getDistributeLBAndStep(OpBuilder &b,
+                                                       Location loc, Value lb,
+                                                       Value step, Value procId,
+                                                       Value nprocs) {
+  AffineExpr s0, s1, s2;
+  bindSymbols(b.getContext(), s0, s1, s2);
+  auto offsetMap = AffineMap::get(0, 3, {s0 + s1 * s2});
+  auto stepMap = AffineMap::get(0, 2, {s0 * s1});
+  Value distributeLB =
+      makeComposedAffineApply(b, loc, offsetMap, ValueRange{lb, procId, step});
+  Value distributeStep =
+      makeComposedAffineApply(b, loc, stepMap, ValueRange{step, nprocs});
+  return {distributeLB, distributeStep};
+}
+
+//===----------------------------------------------------------------------===//
+// TileDispatchUsingSCFForOp pattern implementation.
+//===----------------------------------------------------------------------===//
+
+namespace {
+// Till `scf.for` can take an `OpFoldResult` for lb, ub and step, use of `Range`
+// with `OpFoldResult` causes lots of `OpFoldResult` -> `Value` conversion.
+// This struct is a mirror of `Range` with `Value` type fields.
+struct RangeVal {
+  Value offset;
+  Value size;
+  Value stride;
+};
+}  // namespace
+
+/// Generate an empty loop nest that represents the tiled loop nest shell.
+/// - `loopRanges` specifies the lb, ub and step of the untiled iteration space.
+/// - `tileSizeVals` is the tile sizes to use. Zero represent untiled loops.
+/// - In `offsets` and `sizes` return the multi-dimensional offset and size of
+/// the
+///   tile processed within the inner most loop.
+static SmallVector<scf::ForOp> generateTileLoopNest(
+    OpBuilder &builder, Location loc, ArrayRef<RangeVal> loopRanges,
+    ArrayRef<Value> tileSizeVals,
+    ArrayRef<linalg::DistributionMethod> distributionMethod,
+    ArrayRef<linalg::ProcInfo> procInfo, SmallVector<OpFoldResult> &offsets,
+    SmallVector<OpFoldResult> &sizes) {
```

Comment:
Prefer to use ArrayRef<T> or SmallVectorImpl<T> as a parameter type.

https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h

Reply:
Might not be relevant here (and might be dated). If we use `SmallVector<T>` everywhere, the justification to use of `SmallVectorImpl<T>` goes away...

Reply:
It's true, but there is an assumption that we use `SmallVector<T>` everywhere. We don't ask all the callers to know the assumption before using the method. It's also about style consistency.